Report bad tightening/loosening of attributes in OHI

jcouv opened the pull request:
Implements the last part of https://github.com/dotnet/roslyn/issues/36039 (new enforcement of nullable attributes).

In terms of design:
- we're validating that parameter values can be assigned between overridden and overriding method (or implemented/implementing) accounting for nullable attributes and ref kind
- we're validating that `[Maybe/NotNull]` attributes on by-value/`in` parameters are matching

sharwell commented on docs/compilers/CSharp/Compiler Breaking Changes - post VS2019.md:
```diff
@@ -108,3 +108,16 @@ Similarly, a value from a `[DisallowNull] ref string? p` parameter will be assum
 On the other hand, we'll warn for returning a `null` from a method declared with `[NotNull] string?` return type, and we'll treat the value from a `[AllowNull] ref string p` parameter as maybe-null.
 Conditional attributes are treated leniently. For instance, no warning will be produced for assigning a maybe-null value to an `[MaybeNullWhen(...)] out string p` parameter.
 
+19. https://github.com/dotnet/roslyn/issues/36039 In *Visual Studio 2019 version 16.5* and onwards, the compiler did not check the usage of nullable flow annotation attributes, such as `[MaybeNull]` or `[NotNull]`, in overrides or implementations. In *Visual Studio 2019 version 16.5*, those usages are checked to respect null discipline. For example:
```
```suggestion
19. https://github.com/dotnet/roslyn/issues/36039 Prior to *Visual Studio 2019 version 16.5*, the compiler did not check the usage of nullable flow annotation attributes, such as `[MaybeNull]` or `[NotNull]`, in overrides or implementations. In *Visual Studio 2019 version 16.5* and onwards, those usages are checked to respect null discipline. For example:
```

sharwell commented on docs/features/nullable-reference-types.md:
```diff
@@ -99,7 +99,24 @@ The `Interlocked.CompareExchange` methods have special handling in flow analysis
-The use of any member returning `[MaybeNull]T` for an unconstrained type parameter `T` produces a warning, just like `default(T)` (see below). For instance, `listT.FirstOrDefault();` would produce a warning.
+
+Enforcing nullability attributes within method bodies:
+- An input parameter marked with [AllowNull] is initialized with a maybe-null (or maybe-default) state.
```
```suggestion
- An input parameter marked with `[AllowNull]` is initialized with a maybe-null (or maybe-default) state.
```

sharwell commented on docs/features/nullable-reference-types.md:
```diff
@@ -99,7 +99,24 @@ The `Interlocked.CompareExchange` methods have special handling in flow analysis
+
+Enforcing nullability attributes within method bodies:
+- An input parameter marked with [AllowNull] is initialized with a maybe-null (or maybe-default) state.
+- An input parameter marked with [DisallowNull] is initialized with a not-null state.
```
```suggestion
- An input parameter marked with `[DisallowNull]` is initialized with a not-null state.
```

sharwell commented on docs/features/nullable-reference-types.md:
```diff
@@ -99,7 +99,24 @@ The `Interlocked.CompareExchange` methods have special handling in flow analysis
+Enforcing nullability attributes within method bodies:
+- An input parameter marked with [AllowNull] is initialized with a maybe-null (or maybe-default) state.
+- An input parameter marked with [DisallowNull] is initialized with a not-null state.
+- A parameter marked with [MaybeNull] or [MaybeNullWhen] can be assigned a maybe-null value, without warning. Same for return values. Same for a nullable parameter marked with [NotNullWhen] (the attribute is ignored).
```
```suggestion
- A parameter marked with `[MaybeNull]` or `[MaybeNullWhen]` can be assigned a maybe-null value, without warning. Same for return values. Same for a nullable parameter marked with `[NotNullWhen]` (the attribute is ignored).
```

sharwell commented on docs/features/nullable-reference-types.md:
```diff
@@ -99,7 +99,24 @@ The `Interlocked.CompareExchange` methods have special handling in flow analysis
+- An input parameter marked with [AllowNull] is initialized with a maybe-null (or maybe-default) state.
+- An input parameter marked with [DisallowNull] is initialized with a not-null state.
+- A parameter marked with [MaybeNull] or [MaybeNullWhen] can be assigned a maybe-null value, without warning. Same for return values. Same for a nullable parameter marked with [NotNullWhen] (the attribute is ignored).
+- A parameter marked with [NotNull] will produce a warning when assigned a maybe-null value. Same for return values.
```
This only applies to `out` and `ref` parameters right?

```suggestion
- A by-reference parameter marked with `[NotNull]` will produce a warning when assigned a maybe-null value. Same for return values.
```

jcouv replied:
@sharwell, no this applies to by-value parameters as well. I'll update test `UnconditionalAttributesAffectMethodBodies_ReferenceType` to show that behavior.

---
In reply to: [373548582](https://github.com/dotnet/roslyn/pull/41336#discussion_r373548582) [](ancestors = 373548582)

sharwell commented on docs/features/nullable-reference-types.md:
```diff
@@ -99,7 +99,24 @@ The `Interlocked.CompareExchange` methods have special handling in flow analysis
+- An input parameter marked with [DisallowNull] is initialized with a not-null state.
+- A parameter marked with [MaybeNull] or [MaybeNullWhen] can be assigned a maybe-null value, without warning. Same for return values. Same for a nullable parameter marked with [NotNullWhen] (the attribute is ignored).
+- A parameter marked with [NotNull] will produce a warning when assigned a maybe-null value. Same for return values.
+- The state of a parameter marked with [Maybe/NotNullWhen] is checked upon exiting the method instead.
```
```suggestion
- The state of a parameter marked with `[MaybeNullWhen]`/`[NotNullWhen]` is checked upon exiting the method instead.
```

sharwell commented on docs/features/nullable-reference-types.md:
```diff
@@ -99,7 +99,24 @@ The `Interlocked.CompareExchange` methods have special handling in flow analysis
+- A parameter marked with [MaybeNull] or [MaybeNullWhen] can be assigned a maybe-null value, without warning. Same for return values. Same for a nullable parameter marked with [NotNullWhen] (the attribute is ignored).
+- A parameter marked with [NotNull] will produce a warning when assigned a maybe-null value. Same for return values.
+- The state of a parameter marked with [Maybe/NotNullWhen] is checked upon exiting the method instead.
+- A method marked with [DoesNotReturn] will produce a warning if it returns or exits normally.
```
```suggestion
- A method marked with `[DoesNotReturn]` will produce a warning if it returns or exits normally.
```

sharwell commented on docs/features/nullable-reference-types.md:
```diff
@@ -99,7 +99,24 @@ The `Interlocked.CompareExchange` methods have special handling in flow analysis
+- The state of a parameter marked with [Maybe/NotNullWhen] is checked upon exiting the method instead.
+- A method marked with [DoesNotReturn] will produce a warning if it returns or exits normally.
+
+Note: we don't validate the internal consistency of auto-properties, so it is possible to misused attributes on auto-props as on fields. For example: `[AllowNull, NotNull] public TOpen P { get; set; }`.
```
```suggestion
Note: we don't validate the internal consistency of auto-properties, so it is possible to misuse attributes on auto-props as on fields. For example: `[AllowNull, NotNull] public TOpen P { get; set; }`.
```

sharwell replied:
💡 Consider breaking out a section specifically to point out things that are _not_ checked

sharwell commented on docs/features/nullable-reference-types.md:
```diff
@@ -99,7 +99,24 @@ The `Interlocked.CompareExchange` methods have special handling in flow analysis
+
+Enforcing nullability attributes when overriding and implementing:
+In addition to checking the types in overrides/implementations are compatible with overridden/implemented members, we also check that the nullability attributes are compatible.
+- For input parameters (by-value and `in`), we check that the worst allowed value by the overridden/implemented parameter can be assigned to the overriding/implementing parameter.
```
```suggestion
- For input parameters (by-value and `in`), we check that the widest allowed value by the overridden/implemented parameter can be assigned to the overriding/implementing parameter.
```

sharwell commented on docs/features/nullable-reference-types.md:
```diff
@@ -99,7 +99,24 @@ The `Interlocked.CompareExchange` methods have special handling in flow analysis
+Enforcing nullability attributes when overriding and implementing:
+In addition to checking the types in overrides/implementations are compatible with overridden/implemented members, we also check that the nullability attributes are compatible.
+- For input parameters (by-value and `in`), we check that the worst allowed value by the overridden/implemented parameter can be assigned to the overriding/implementing parameter.
+- For output parameters (`out` and return values), we check that the worst produced value by the overriding/implementing parameter can be assigned to the overridden/implemented parameter.
```
```suggestion
- For output parameters (`out` and return values), we check that the widest produced value by the overriding/implementing parameter can be assigned to the overridden/implemented parameter.
```

sharwell commented on docs/features/nullable-reference-types.md:
```diff
@@ -99,7 +99,24 @@ The `Interlocked.CompareExchange` methods have special handling in flow analysis
+- For input parameters (by-value and `in`), we check that the worst allowed value by the overridden/implemented parameter can be assigned to the overriding/implementing parameter.
+- For output parameters (`out` and return values), we check that the worst produced value by the overriding/implementing parameter can be assigned to the overridden/implemented parameter.
+- For `ref` parameters and return values, we check both.
+- We check that the post-condition contract `[Not/MaybeNull]` on input parameters is enforced by overriding/implementing members.
```
```suggestion
- We check that the post-condition contract `[NotNull]`/`[MaybeNull]` on input parameters is present and enforced by overriding/implementing members.
```

sharwell commented on docs/features/nullable-reference-types.md:
```diff
@@ -99,7 +99,24 @@ The `Interlocked.CompareExchange` methods have special handling in flow analysis
+- For output parameters (`out` and return values), we check that the worst produced value by the overriding/implementing parameter can be assigned to the overridden/implemented parameter.
+- For `ref` parameters and return values, we check both.
+- We check that the post-condition contract `[Not/MaybeNull]` on input parameters is enforced by overriding/implementing members.
+- We check that overrides/implementations have the `[DoesNotReturn]` attribute if the overridden/implemented member has it.
```
❔ Do we allow an implementation to use `[DoesNotReturn]` if the original definition used `[DoesNotReturnIf]`?

jcouv replied:
I'll add a test, but I think we allow it and that behavior is correct. 
The implementation abides by the overridden/implemented method's contract.

---
In reply to: [373550920](https://github.com/dotnet/roslyn/pull/41336#discussion_r373550920) [](ancestors = 373550920)

jcouv replied:
It turns out the problem doesn't even present itself. `[DoesNotReturn]` is applied to methods, but `[DoesNotReturnIf(...)]` is applied to parameters. They don't cross paths.

cston commented on docs/features/nullable-reference-types.md:
```diff
@@ -99,7 +99,24 @@ The `Interlocked.CompareExchange` methods have special handling in flow analysis
+- An input parameter marked with `[DisallowNull]` is initialized with a not-null state.
+- A parameter marked with `[MaybeNull]` or `[MaybeNullWhen]` can be assigned a maybe-null value, without warning. Same for return values. Same for a nullable parameter marked with `[NotNullWhen]` (the attribute is ignored).
+- A parameter marked with `[NotNull]` will produce a warning when assigned a maybe-null value. Same for return values.
+- The state of a parameter marked with `[MaybeNullWhen]`\`[NotNullWhen]` is checked upon exiting the method instead.
```
>\ [](start = 56, length = 1)

"/" or "or"

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1157,6 +1160,124 @@ static bool isMaybeDefaultValue(TypeWithState valueType)
+                bool overriddenHasNotNull = (overriddenAnnotations & FlowAnalysisAnnotations.NotNull) == FlowAnalysisAnnotations.NotNull;
+                if (overriddenHasNotNull && !overridingHasNotNull && !forRef)
+                {
+                    // Overriding doesn't conform to contract of overridden (ie. promise not to return if parameter is null)
```
>promise not to return if parameter is null [](start = 81, length = 42)

What does "promise not to return" mean for parameters?

jcouv replied:
`void AssertNotNull<T>([NotNull] T t)` would be an example. The method promises not to return if the parameter is null.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1157,6 +1160,124 @@ static bool isMaybeDefaultValue(TypeWithState valueType)
+                {
+                    // Can't assign value from overriding to overridden in true case
+                    return false;
+                }
```
Consider extracting a helper method, with `sense` parameter, for use here and `false` case below.

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs:
```diff
@@ -1050,37 +1053,44 @@ static bool isOrContainsErrorType(TypeSymbol typeSymbol)
-                                                     }
-                :
-                                                     (Action<DiagnosticBag, MethodSymbol, MethodSymbol, ParameterSymbol, Location>)null,
+                                                 checkReturnType ? reportBadReturn : (ReportMismatchinReturnType<Location>)null,
```
>(ReportMismatchinReturnType<Location>) [](start = 85, length = 38)

Are the casts needed here or for `checkParameters`?

jcouv replied:
Yes, the casts are necessary.

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs:
```diff
@@ -1050,37 +1053,44 @@ static bool isOrContainsErrorType(TypeSymbol typeSymbol)
-                                                     }
-                :
-                                                     (Action<DiagnosticBag, MethodSymbol, MethodSymbol, ParameterSymbol, Location>)null,
+                                                 checkReturnType ? reportBadReturn : (ReportMismatchinReturnType<Location>)null,
```
>reportBadReturn [](start = 67, length = 15)

The local functions will result in delegate instances. Consider inlining the local functions instead.

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs:
```diff
@@ -1104,14 +1126,25 @@ static bool isOrContainsErrorType(TypeSymbol typeSymbol)
-                    reportMismatchInParameterType(diagnostics, overriddenMethod, overridingMethod, overridingParameters[i], extraArgument);
+                    reportMismatchInParameterType(diagnostics, overriddenMethod, overridingMethod, parameter, false, extraArgument);
+                }
+                else if (reportMismatchInParameterType != null &&
```
>reportMismatchInParameterType != null [](start = 25, length = 37)

Already checked before the `for` loop.

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs:
```diff
@@ -1104,14 +1126,25 @@ static bool isOrContainsErrorType(TypeSymbol typeSymbol)
             for (int i = 0; i < overriddenMethod.ParameterCount; i++)
             {
                 var overriddenParameterType = overriddenParameters[i].TypeWithAnnotations;
```
>overriddenParameters[i] [](start = 46, length = 23)

Consider extracting a local.

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs:
```diff
@@ -1050,37 +1053,44 @@ static bool isOrContainsErrorType(TypeSymbol typeSymbol)
+            }
+
             var conversions = compilation.Conversions.WithNullability(true);
             if (reportMismatchInReturnType != null &&
```
>reportMismatchInReturnType != null [](start = 16, length = 34)

Please consider checking this condition once, around this case and the one below.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1157,6 +1160,124 @@ static bool isMaybeDefaultValue(TypeWithState valueType)
+                if (overriddenHasNotNull && !overridingHasNotNull && !forRef)
+                {
+                    // Overriding doesn't conform to contract of overridden (ie. promise not to return if parameter is null)
+                    return false;
```
>false [](start = 27, length = 5)

Do we get here with substituted types?
```
abstract class A<T>
{
    [return: NotNull] abstract T F();
}
class B : A<string?>
{
    override string F() => string.Empty;
}
```

jcouv replied:
We only get here with input parameters. `[NotNull]` takes a special meaning on inputs, namely "the method will not return if the input is null". This contract can only be represented with the `[NotNull]` attribute and cannot be represented with types. So we need the attribute regardless of substitution.

I'll double-check the behavior for `ref` scenarios. The `!forRef` check might be wrong.

cston commented:
>         public void AllowNull_OHI_OnInterface()

Consider dropping the "_OHI" qualifier from tests if it's not needed.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:20882 in 3b9d591. [](commit_id = 3b9d5915f12ea63814308c2e864a023214cf6c1f, deletion_comment = False)

cston commented:
>             // Note: warnings on F4 aren't great, but are probably deserved

There are no warnings on F4.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:29419 in 3b9d591. [](commit_id = 3b9d5915f12ea63814308c2e864a023214cf6c1f, deletion_comment = False)

cston commented:
>     public override void M() { } // 1

Consider using `=> throw null;` so it's clear the body is ignored for the warning.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:30305 in 3b9d591. [](commit_id = 3b9d5915f12ea63814308c2e864a023214cf6c1f, deletion_comment = False)

cston commented:
> }

Are we testing `[DoesNotReturn]` on derived or implementing method only?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:30326 in 3b9d591. [](commit_id = 3b9d5915f12ea63814308c2e864a023214cf6c1f, deletion_comment = False)

cston commented:
> }

Include implicit implementations as well.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:30326 in 3b9d591. [](commit_id = 3b9d5915f12ea63814308c2e864a023214cf6c1f, deletion_comment = False)

jcouv commented:
> Are we testing [DoesNotReturn] on derived or implementing method only?

`DoesNotReturn_OnOverriddenMethod` covers the override scenario

cston commented:
> ";

Are we testing missing attribute on implicit implementation?
```
interface I
{
   [DoesNotReturn] void F();
}
class C : I
{
    public void F() => throw null;
}
```


---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:25823 in 64124a2. [](commit_id = 64124a24336264511fa311d3235ccd3f7cdcddbf, deletion_comment = False)

jcouv commented:
@dotnet/roslyn-compiler for a second review. Thanks

jcouv commented:
Dogfooding this PR on the `runtime` repo revealed a first wave of warnings. I'll investigate and iterate.

``` 
P:\runtime\src\coreclr\coreclr.proj(25,5): error MSB3073: The command ""P:\runtime\src\coreclr\build.cmd" -skiptests -debug" exited with code 1.
P:\runtime\src\mono\mono.proj(58,5): error MSB3073: The command "msvc\run-msbuild.bat build x64 Debug sgen "/p:MONO_BUILD_DIR_PREFIX=""P:\runtime\artifacts\obj\mono/Windows_NT.x64.Debug/"" /p:MONO_ENABLE_NETCORE=true /p:CL_MPCount=8 /v:minimal" msvc\mono-netcore.sln" exited with code 1.
src\System\TypeSpec.cs(593,57): error CS8767: Nullability of reference types in type of parameter 'other' of 'bool ATypeName.Equals(TypeName other)' doesn't match implicitly implemented member 'bool IEquatable<TypeName>.Equals(TypeName other)' because of nullability attributes. [P:\runtime\src\mono\netcore\System.Private.CoreLib\System.Private.CoreLib.csproj]
src\System\TypeIdentifier.cs(131,48): error CS8767: Nullability of reference types in type of parameter 'other' of 'bool ATypeName.Equals(TypeName other)' doesn't match implicitly implemented member 'bool IEquatable<TypeName>.Equals(TypeName other)' because of nullability attributes. [P:\runtime\src\mono\netcore\System.Private.CoreLib\System.Private.CoreLib.csproj]
src\System\TypeIdentifier.cs(165,49): error CS8767: Nullability of reference types in type of parameter 'other' of 'bool ATypeName.Equals(TypeName other)' doesn't match implicitly implemented member 'bool IEquatable<TypeName>.Equals(TypeName other)' because of nullability attributes. [P:\runtime\src\mono\netcore\System.Private.CoreLib\System.Private.CoreLib.csproj]
src\System\TypeIdentifier.cs(204,49): error CS8767: Nullability of reference types in type of parameter 'other' of 'bool ATypeName.Equals(TypeName other)' doesn't match implicitly implemented member 'bool IEquatable<TypeName>.Equals(TypeName other)' because of nullability attributes. [P:\runtime\src\mono\netcore\System.Private.CoreLib\System.Private.CoreLib.csproj]
P:\runtime\src\libraries\System.Private.CoreLib\src\System\SpanHelpers.BinarySearch.cs(78,24): error CS8767: Nullability of reference types in type of parameter 'other' of 'int ComparerComparable<T, TComparer>.CompareTo(T other)' doesn't match implicitly implemented member 'int IComparable<T>.CompareTo(T other)' because of nullability attributes. [P:\runtime\src\mono\netcore\System.Private.CoreLib\System.Private.CoreLib.csproj]
src\System\TypeIdentifier.cs(71,16): error CS8767: Nullability of reference types in type of parameter 'other' of 'bool ATypeName.Equals(TypeName other)' doesn't match implicitly implemented member 'bool IEquatable<TypeName>.Equals(TypeName other)' because of nullability attributes. [P:\runtime\src\mono\netcore\System.Private.CoreLib\System.Private.CoreLib.csproj]
P:\runtime\src\libraries\System.Private.CoreLib\src\System\Collections\Generic\Comparer.cs(46,29): error CS8765: Type of parameter 'x' doesn't match overridden member because of nullability attributes. [P:\runtime\src\mono\netcore\System.Private.CoreLib\System.Private.CoreLib.csproj]
P:\runtime\src\libraries\System.Private.CoreLib\src\System\Collections\Generic\Comparer.cs(46,29): error CS8765: Type of parameter 'y' doesn't match overridden member because of nullability attributes. [P:\runtime\src\mono\netcore\System.Private.CoreLib\System.Private.CoreLib.csproj]
P:\runtime\src\libraries\System.Private.CoreLib\src\System\Diagnostics\Debug.cs(119,51): error CS8763: A method marked [DoesNotReturn] should not return. [P:\runtime\src\mono\netcore\System.Private.CoreLib\System.Private.CoreLib.csproj]
P:\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\InteropServices\MemoryMarshal.cs(310,13): error CS8762: Parameter 'manager' may not have a null value when exiting with 'true'. [P:\runtime\src\mono\netcore\System.Private.CoreLib\System.Private.CoreLib.csproj]
P:\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\InteropServices\MemoryMarshal.cs(338,13): error CS8762: Parameter 'manager' may not have a null value when exiting with 'true'. [P:\runtime\src\mono\netcore\System.Private.CoreLib\System.Private.CoreLib.csproj]
P:\runtime\src\libraries\System.Private.CoreLib\src\System\HashCode.cs(306,57): error CS8607: A possible null value may not be used for a type marked with [NotNull] or [DisallowNull] [P:\runtime\src\mono\netcore\System.Private.CoreLib\System.Private.CoreLib.csproj]
System.Runtime.cs(4216,52): error CS8763: A method marked [DoesNotReturn] should not return. [P:\runtime\src\libraries\System.Runtime\ref\System.Runtime.csproj]
System.Runtime.cs(4219,75): error CS8763: A method marked [DoesNotReturn] should not return. [P:\runtime\src\libraries\System.Runtime\ref\System.Runtime.csproj]
System.Runtime.cs(7450,31): error CS8763: A method marked [DoesNotReturn] should not return. [P:\runtime\src\libraries\System.Runtime\ref\System.Runtime.csproj]
System.Runtime.cs(7452,61): error CS8763: A method marked [DoesNotReturn] should not return. [P:\runtime\src\libraries\System.Runtime\ref\System.Runtime.csproj]
```

To repro this locally, run `build.cmd -bl` on the `runtime` repo after adding a custom version number for toolset compiler (in `Versions.props`) and a custom nuget folder in `nuget.json`.

RikkiGibson commented:
>     public virtual void F1<T>([AllowNull]T  t1, [AllowNull] out T  t2, [AllowNull] ref T  t3, [AllowNull] in T  t4)                  => throw null!;

It feels wrong that we ever allow input-affecting attributes on output parameters.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:29427 in 4ab452c. [](commit_id = 4ab452c43e6d2725f85f114c2a460a6fd411b16b, deletion_comment = False)

RikkiGibson commented:
>             // Note: we're missing warnings on F1 because the overridden method with substituted T from Derived has an oblivious T

I don't understand how this is the case, the whole file is in a nullable enable context.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:30581 in 4ab452c. [](commit_id = 4ab452c43e6d2725f85f114c2a460a6fd411b16b, deletion_comment = False)

RikkiGibson commented:
>     public virtual bool F6<T>([MaybeNullWhen(true)]T  t1, [MaybeNullWhen(true)] out T  t2, [MaybeNullWhen(true)] ref T  t3, [MaybeNullWhen(true)] in T  t4) where T : notnull => throw null!;

Are we intentionally missing an override for F6?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:30644 in 4ab452c. [](commit_id = 4ab452c43e6d2725f85f114c2a460a6fd411b16b, deletion_comment = False)

RikkiGibson commented:
>                 // (12,23): warning CS8764: Return type doesn't match overridden member because of nullability attributes.

It feels like these warning messages could reasonably be divided between "An overriding (output parameter|return) may be null when the base (parameter|return) cannot be null" and "An overriding input parameter doesn't accept null but the base parameter does accept null".

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:35324 in 4ab452c. [](commit_id = 4ab452c43e6d2725f85f114c2a460a6fd411b16b, deletion_comment = False)

RikkiGibson commented:
>             // Note: we're missing the warning on F6 because the overridden method with substituted T from Derived has an oblivious T

oh, is it oblivious because we haven't performed nullability analysis at the time we get the overridden symbol?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:35413 in 4ab452c. [](commit_id = 4ab452c43e6d2725f85f114c2a460a6fd411b16b, deletion_comment = False)

RikkiGibson commented:
>     [return: NotNull] public override T F1<T>(T t) => t; // 1

I was thrown off by this test because most of the others have the base first and then the derived. Consider reordering the code for uniformity with the other tests.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:35678 in 4ab452c. [](commit_id = 4ab452c43e6d2725f85f114c2a460a6fd411b16b, deletion_comment = False)

RikkiGibson commented:
>     public bool TryGetValue([MaybeNullWhen(false)] out T t) => throw null!;

I would expect that if we had e.g. `[NotNullWhen(true)]`, there would be no warning. Is that the case? Any existing test that shows this?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:30521 in 4ab452c. [](commit_id = 4ab452c43e6d2725f85f114c2a460a6fd411b16b, deletion_comment = False)

RikkiGibson commented:
>     public override void F1<T>(T  t1, out T  t2, ref T  t3, in T  t4)                  => throw null!; // t1, t4

Looking at `T t1` specifically: does it make sense for users to write a method which "introduces doubt" about the null state in the way that `x != null` or `x?.y` checks do?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:30429 in 4ab452c. [](commit_id = 4ab452c43e6d2725f85f114c2a460a6fd411b16b, deletion_comment = False)

RikkiGibson commented:
>     [DoesNotReturn]

Are there any tests for hiding members on a derived class?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:30322 in 4ab452c. [](commit_id = 4ab452c43e6d2725f85f114c2a460a6fd411b16b, deletion_comment = False)

RikkiGibson commented:
>     public override bool F3<T>(T? t1, out T? t2, ref T? t3, in T? t4) where T : class  => throw null!; // t2, t3

Does `[NotNullWhen]` just not do anything by design for `None/In` parameters? I would have expected a similar behavior to `[NotNull]` where we make a claim that the argument is not-null when the method returns a certain value.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:29650 in 4ab452c. [](commit_id = 4ab452c43e6d2725f85f114c2a460a6fd411b16b, deletion_comment = False)

RikkiGibson commented:
>     public override void F1<T>(T  t1, out T  t2, ref T  t3, in T  t4)                  => throw null!; // t1, t3, t4

Should these trailing comments be deleted?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:29523 in 4ab452c. [](commit_id = 4ab452c43e6d2725f85f114c2a460a6fd411b16b, deletion_comment = False)

RikkiGibson commented:
>     public override bool F1<T>(T  t1, out T  t2, ref T  t3, in T  t4)                  => throw null!; // t1, t4

Should these trailing comments be deleted?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:30495 in 4ab452c. [](commit_id = 4ab452c43e6d2725f85f114c2a460a6fd411b16b, deletion_comment = False)

RikkiGibson commented:
>     public override void F1<T>([AllowNull]T  t1, [AllowNull] out T  t2, [AllowNull] ref T  t3, [AllowNull] in T  t4)                  => throw null!;

Any need to test this scenario for `T : notnull`?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:29547 in 4ab452c. [](commit_id = 4ab452c43e6d2725f85f114c2a460a6fd411b16b, deletion_comment = False)

RikkiGibson commented:
>     public virtual void F1<T>([AllowNull]T  t1, [AllowNull] out T  t2, [AllowNull] ref T  t3, [AllowNull] in T  t4)                  => throw null!;

Is use of nullability attributes outside of a nullable context something we intend to support? Or are we just recording the existing behavior?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:29473 in 4ab452c. [](commit_id = 4ab452c43e6d2725f85f114c2a460a6fd411b16b, deletion_comment = False)

RikkiGibson commented:
>     [MaybeNull] public override string P2 { set { throw null!; } } // 0

I don't see this diagnostic in VerifyDiagnostics

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:21475 in 4ab452c. [](commit_id = 4ab452c43e6d2725f85f114c2a460a6fd411b16b, deletion_comment = False)

RikkiGibson commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1157,6 +1160,121 @@ static bool isMaybeDefaultValue(TypeWithState valueType)
+                    overridingType.ToTypeWithState(),
+                    makeUnconditionalAnnotation(overridingAnnotations, sense));
+
+                var destAnnotationsWhenTrue = ToInwardAnnotations(makeUnconditionalAnnotation(overriddenAnnotations, sense));
```
>ToInwardAnnotations [](start = 46, length = 19)

What does `ToInwardAnnotations` mean? Does it refer to annotations that affect input, e.g. `AllowNull`? Maybe it would be better to use the term `inbound` or `input` as it has had other uses throughout the walker and we can avoid using multiple terms to refer to the same specific thing.

jcouv replied:
If you have a `void M([AllowNull] string x)` method, inside the body `NullableWalker` needs to see the inwards annotation of `x`, ie. `[MaybeNull]`.
That way the initial state of `x` will be maybe-null (since we allow `null` from the caller).
"Inward" is a different concept, to refer to this reversal.

RikkiGibson commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1157,6 +1160,121 @@ static bool isMaybeDefaultValue(TypeWithState valueType)
+            if (refKind == RefKind.Ref)
+            {
+                // ref variables are invariant
+                return AreParameterAnnotationsCompatible(RefKind.None, overriddenType, overriddenAnnotations, overridingType, overridingAnnotations, forRef: true) &&
```
Why do we pass in `RefKind.None` and also `forRef: true`? What distinguishes the meaning of those parameters?

jcouv replied:
On a `RefKind.None` parameter, `[NotNull]` takes a special meaning. We don't want that meaning to apply to a `ref` argument when it is analyzed as `RefKind.None` + `RefKind.Out`.

RikkiGibson commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -1157,6 +1160,121 @@ static bool isMaybeDefaultValue(TypeWithState valueType)
             }
         }
 
+        internal static bool AreParameterAnnotationsCompatible(
```
Perhaps this should be named `CheckOverridingParameterAnnotations`? Since we tend to "blame" the overriding method for not conforming to the requirements of the overridden method? Also wondering if it might make sense to declare closer to its use site in SourceMemberContainerSymbol_ImplementationChecks.

jcouv replied:
I initially has this method in `SourceMemberContainerSymbol_ImplementationChecks`, but that required to expose multiple private methods of `NulllableWalker` as internal. So I figured this was a more natural place.
I'll adjust the name

jcouv replied:
I ended up keeping the current name because this method doesn't report any diagnostics. The caller is assigning blame to the overriding/implementing parameter ;-)

---
In reply to: [377404557](https://github.com/dotnet/roslyn/pull/41336#discussion_r377404557) [](ancestors = 377404557)

RikkiGibson commented:
Done with review pass (iteration 7)

jcouv commented:
Created PR https://github.com/dotnet/runtime/pull/32090 for impact of this change on `runtime` repo. Tagging @jaredpar as FYI

jcouv commented:
> Is use of nullability attributes outside of a nullable context something we intend to support? Or are we just recording the existing behavior?

We are just recording the existing behavior.

jcouv commented:
> Does [NotNullWhen] just not do anything by design for None/In parameters? I would have expected a similar behavior to [NotNull] where we make a claim that the argument is not-null when the method returns a certain value.

That would make sense, but I'm not sure I have a compelling example.
Filed https://github.com/dotnet/roslyn/issues/41569

jcouv commented:
> Looking at T t1 specifically: does it make sense for users to write a method which "introduces doubt" about the null state in the way that x != null or x?.y checks do?

A `void M([MaybeNull] ...)` method does that. In `M(expr)` we will learn that the expression may be null, just as we do in `expr == null`.

``` csharp
using System.Diagnostics.CodeAnalysis;
#nullable enable

public class C 
{
    public void M([MaybeNull] string x) 
    {
    }
    
    public void M2(string s)
    {
       M(s);
       s.ToString(); //warning CS8602: Dereference of a possibly null reference.
    }
}
```
[sharplab](https://sharplab.io/#v2:CYLg1APgAgDABFAjAOgCIEsCGBzAdgewGcAXdAY0OQGF9gBTAQV0wBsBPQ9QgbgFgAoAMS4ArixaYARizpw6zaXQECoAZgQAmOFTgCA3gLhGE6qABY4AWQAUAbUuY2kugDkxLALoJE8AB4BKXX5jOANg4wBfQ2NoozUEC0sNayR4Qn9Y0MzjG3S+cJC4SgAVfABlYgAndFxsa39uOAB6JoB3TErcGuxtMoAOADYYDRA4VDpKugAzCfkyWXwpuEw4AAciTmk2OFFxOEmZydx55Eyo/gigA===)

jcouv commented:
> `public bool TryGetValue([MaybeNullWhen(false)] out T t) => throw null!;`
> I would expect that if we had e.g. [NotNullWhen(true)], there would be no warning. Is that the case? Any existing test that shows this?

I think that's covered by `NotNullWhenTrue_EnforceInMethodBody` and `NotNullWhenTrue_EnforceInMethodBody_Warn`.
Let me know if those are not the scenarios you had in mind.

jcouv commented:
> oh, is it oblivious because we haven't performed nullability analysis at the time we get the overridden symbol?

It's oblivious because we have a bug in type substitution. Fixing that bug introduces a difficult cycle. Filed a follow-up issue. Details are in https://github.com/dotnet/roslyn/issues/41368

jcouv commented:
Addressed the following comments:

> - I don't see this diagnostic in VerifyDiagnostics
> - Should these trailing comments be deleted?
> - Consider reordering the code for uniformity with the other tests.
> - Are we intentionally missing an override for F6?
> - Are there any tests for hiding members on a derived class?

Fixed/added. Thanks

> I think these warnings are pretty serviceable as is, just something to consider.

Leaving as-is for now. We can tweak diagnostics as we go.

jcouv commented:
This PR flagged more issues in bootstrapping that were introduced in `master`. I'll merge master into this PR and address.

cston commented on src/Compilers/Core/Portable/SourceFileResolver.cs:
```diff
@@ -135,8 +135,10 @@ public override bool Equals(object? obj)
-        public bool Equals(SourceFileResolver other)
+        public bool Equals(SourceFileResolver? other)
         {
+            RoslynDebug.Assert(other is object);
```
>RoslynDebug.Assert(other is object) [](start = 12, length = 35)

This is a `public` method so we should handle `null`.

jcouv replied:
Thanks, indeed. I'll add a null check (since implementation below would crash on dereference...)

---
In reply to: [377720266](https://github.com/dotnet/roslyn/pull/41336#discussion_r377720266) [](ancestors = 377720266)

cston commented on src/Compilers/Core/Portable/PEWriter/MemberRefComparer.cs:
```diff
@@ -18,12 +19,13 @@ internal MemberRefComparer(MetadataWriter metadataWriter)
         }
 
-        public bool Equals(ITypeMemberReference x, ITypeMemberReference y)
+        public bool Equals([AllowNull] ITypeMemberReference x, [AllowNull] ITypeMemberReference y)
```
>[AllowNull] ITypeMemberReference [](start = 27, length = 32)

Consider using `ITypeMemberReference?` rather than `[AllowNull]`.

cston commented on src/Compilers/Core/Portable/PEWriter/MethodSpecComparer.cs:
```diff
@@ -18,12 +19,13 @@ internal MethodSpecComparer(MetadataWriter metadataWriter)
         }
 
-        public bool Equals(IGenericMethodInstanceReference x, IGenericMethodInstanceReference y)
+        public bool Equals([AllowNull] IGenericMethodInstanceReference x, [AllowNull] IGenericMethodInstanceReference y)
```
>[AllowNull] IGenericMethodInstanceReference [](start = 27, length = 43)

`IGenericMethodInstanceReference?`

cston commented on src/Compilers/Core/Portable/PEWriter/TypeSpecComparer.cs:
```diff
@@ -17,7 +18,7 @@ internal TypeSpecComparer(MetadataWriter metadataWriter)
         }
 
-        public bool Equals(ITypeReference x, ITypeReference y)
+        public bool Equals([AllowNull] ITypeReference x, [AllowNull] ITypeReference y)
```
>[AllowNull] ITypeReference [](start = 27, length = 26)

`ITypeReference?`

RikkiGibson commented on Compilers.sln:
```diff
@@ -1,6 +1,6 @@
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 15
-VisualStudioVersion = 15.0.27102.0
+# Visual Studio Version 16
```
intended?

jcouv replied:
Thanks for catching that. My `compilers.sln` keeps getting updated recently and it's been error-prone to avoid adding it... :(
